SO-5842: core API changes to support resource agnostic upgrade mechanism

cmark opened the pull request.

cmark committed:
feat(upgrade): support `dependencies_upgrades` expansion

Expand option key name is WIP.

cmark committed:
test(core): rename assertCodeSystemCreated helper to createCodeSystem

cmark committed:
feat(core): add isSpecial method to ResourceURI

cmark committed:
test: set ES docker image to the latest version

cmark committed:
fix: organize imports in BaseResourceSearchRequest

cmark committed:
refactor(core): remove old/deprecated API of CodeSystem Upgrade

cmark committed:
feat(core): add Resource.get() method to extract custom property...

...from representation

cmark committed:
feat(core): define TerminologyResource.getDependency(String, String)...

...helper method

cmark committed:
feat(core): add toString and dependOnSameResource methods to Dependency

cmark committed:
feat(core): support hidden resources

cmark committed:
feat(core): support hidden resource filtering in Java API

cmark committed:
fix(core): remove duplicate typeRank constant

cmark committed:
fix(api): remove obsolete upgrade representation type

cmark committed:
feat(commons): make OptionsBuilder smarter by adding helper...

...from SearchResourceRequestBuilder

Allow overriding filter options before executing the search request.

cmark committed:
test: add LoggingTestWatcher helper

cmark committed:
feat(core): support searching resources by id prefix

cmark committed:
test(upgrade): remove SnomedExtensionUpgradeTest

apeteri commented on commons/com.b2international.commons/src/com/b2international/commons/options/OptionsBuilder.java:
```diff
@@ -57,12 +74,35 @@ public OptionsBuilder putAll(Map<String, Object> source) {
+		return add(key.name(), value);
+	}
+	
+	public OptionsBuilder add(String key, Object value) {
```
Given this method name, I'd somewhat expect the following capabilities as well:

- append elements to an existing key where the value is an `Iterable`
- "upgrade" from single element to `Iterable` where the associated value is not an `Iterable`

Could this functionality be folded entirely into `put(String, Object)` as an alternative?

cmark replied:
Not sure yet, I'll think about it and will resolve it next week on a separate PR.

cmark replied:
Merged the add method into the put(String, Object) method, see here: https://github.com/b2ihealthcare/snow-owl/commit/f6d19225eac65a24faa7fd01dbe9c56a2c6baf8e

cmark committed:
fix(core): backward compatibility of hidden filter on pre-9.x indices

cmark committed:
feat(core): support pluggable special path segment resolution...

...when resolving a resource uri to an actual branch path.

Prevent using special URI paths in dependencies.

cmark committed:
chore(test): export snomed test packages

cmark committed:
refactor: remove deprecated upgrade complete API

apeteri committed:
fix(core.rest.tests): Add test method from 9.x to resolve merge conflict

apeteri committed:
Merge branch '9.x' into feature/SO-5842-resource-agnostic-upgrades

apeteri reviewed (approved):
🍏

nagyo commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/TerminologyResource.java:
```diff
@@ -125,42 +130,71 @@ public static abstract class Expand extends Resource.Expand {
-	
+
+	/**
+	 * Searches the dependency array for the dependency URI that has the matching scope. If there is no such dependency can be found in the dependency
```
```suggestion
	 * Searches the dependency array for the dependency URI that has the matching scope. If there is no such dependency in the dependency
```

nagyo commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/TerminologyResource.java:
```diff
@@ -125,42 +130,71 @@ public static abstract class Expand extends Resource.Expand {
+	/**
+	 * Searches the dependency array for the dependency URI that has the matching scope. If there is no such dependency can be found in the dependency
+	 * array then the system takes the current settings object into account and tries to find the appropriate settingsKeyed value as fallback
+	 * dependency URI. If none of them can be found this method return <code>null</code>.
```
```suggestion
	 * dependency URI. If none of them can be found this method returns <code>null</code>.
```

nagyo commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/TerminologyResource.java:
```diff
@@ -317,18 +355,19 @@ public Builder toDocumentBuilder() {
-				.oid(getOid())
 				.settings(getSettings())
+				// terminology resource fields
+				.branchPath(getBranchPath()).oid(getOid())
```
```suggestion
				.branchPath(getBranchPath())
				.oid(getOid())
```

nagyo commented on core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseMetadataResourceConverter.java:
```diff
@@ -13,127 +13,125 @@
-				});
+	
+	protected final void expandDependencyUpgrades(List<R> results) {
+		if (!expand().containsKey("dependencies_upgrades")) {
```
I thought that we are following java naming conventions for expand keywords? E.g. `inboundRelationships`, `statedDescendants`, `referenceSet`. The wording should probably follow the method's name or the other way around, e.g. `dependencyUpgrades`, `upgradeDependencies`

nagyo replied:
It would also be nice to save this in a static field somewhere, like:
https://github.com/b2ihealthcare/snow-owl/blob/4437ec2b51e6850a1092ed1fec5650b529b9bfa4/snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/domain/SnomedConcept.java#L113

cmark replied:
This is a temporary expand key, we need dot notation support here, like this `dependencies.upgrades()` but unfortunately the current expand parser does not like dots in property names and fails to parse the object. A minor refactor is required to make it work so we had to use underscore instead of dot for the separator.

nagyo commented on snomed/com.b2international.snowowl.snomed.core.rest.tests/src/com/b2international/snowowl/snomed/core/rest/ext/SnomedComponentEffectiveTimeRestoreTest.java:
```diff
@@ -129,102 +124,4 @@ public void restoreInternationalEffectiveTimeOnExtension() throws Exception {
-		concept = getConcept(upgradeCodeSystem.getResourceURI(), moduleId);
-		assertEquals(effectiveTime, concept.getEffectiveTime());
-	}
-	
```
Please don't forget to add back or refactor these unit tests later on.

cmark replied:
Yup, that's on my to-do list.

nagyo commented on snomed/com.b2international.snowowl.snomed.core.rest.tests/src/com/b2international/snowowl/snomed/core/rest/ext/SnomedExtensionUpgradeTest.java:
```diff
@@ -1,1517 +0,0 @@
-		return extensionConcept.getDescriptions().getItems().stream().filter(d -> d.getTerm().equals(extensionFsnTerm)).findFirst().get();
-	}
-	
-}
```
Same for these

nagyo reviewed (approved):
LGTM with some minor notes that should be addressed later

nagyo committed:
Update core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/TerminologyResource.java

nagyo committed:
Update core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/TerminologyResource.java

nagyo committed:
Update core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/TerminologyResource.java

nagyo committed:
Merge remote-tracking branch 'origin/9.x' into feature/SO-5842-resource-agnostic-upgrades

nagyo committed:
chore(releng): add snomed.core.rest.tests to update site
